src/Storages/StorageDelta.cpp

```diff
@@ -0,0 +1,289 @@
+        /* max single read retries */ 10,
+        ReadSettings{});
+}
+
+void JsonMetadataGetter::handleJSON(const JSON & json)
+{
+    if (json.has("add"))
+    {
+        auto path = json["add"]["path"].getString();
+        auto timestamp = json["add"]["modificationTime"].getInt();
+
+        metadata.setLastModifiedTime(path, timestamp);
+    }
+    else if (json.has("remove"))
+    {
+        auto path = json["remove"]["path"].getString();
+        auto timestamp = json["remove"]["deletionTimestamp"].getInt();
+
+        metadata.remove(path, timestamp);
+    }
+}
+
+StorageDelta::StorageDelta(
+    const S3::URI & uri_,
+    const String & access_key_,
+    const String & secret_access_key_,
+    const StorageID & table_id_,
+    const String & format_name_,
+    ColumnsDescription columns_,
+    const ConstraintsDescription & constraints_,
+    const String & comment,
+    ContextPtr context_)
+    : IStorage(table_id_)
+    , base_configuration({uri_, access_key_, secret_access_key_, {}, {}, {}})
+    , log(&Poco::Logger::get("StorageDeltaLake (" + table_id_.table_name + ")"))
+    , table_path(uri_.key)
+{
+    StorageInMemoryMetadata storage_metadata;
+    StorageS3::updateS3Configuration(context_, base_configuration);
+
+    JsonMetadataGetter getter{base_configuration, table_path};
+
+    auto keys = getter.getFiles();
+
+    auto new_uri = base_configuration.uri.uri.toString() + generateQueryFromKeys(std::move(keys));
+
+    LOG_DEBUG(log, "New uri: {}", new_uri);
+    LOG_DEBUG(log, "Table path: {}", table_path);
+    auto s3_uri = S3::URI(Poco::URI(new_uri));
+
+    if (columns_.empty())
+    {
+        columns_ = StorageS3::getTableStructureFromData(format_name_, s3_uri, access_key_, secret_access_key_, "", false, {}, context_);
+        storage_metadata.setColumns(columns_);
+    }
+    else
+        storage_metadata.setColumns(columns_);
+
+    storage_metadata.setConstraints(constraints_);
+    storage_metadata.setComment(comment);
+    setInMemoryMetadata(storage_metadata);
+
+    s3engine = std::make_shared<StorageS3>(
+        s3_uri,
+        access_key_,
+        secret_access_key_,
+        table_id_,
+        format_name_,
+        base_configuration.rw_settings,
+        columns_,
+        constraints_,
+        comment,
+        context_,
+        std::nullopt);
+}
+
+Pipe StorageDelta::read(
+    const Names & column_names,
+    const StorageSnapshotPtr & storage_snapshot,
+    SelectQueryInfo & query_info,
+    ContextPtr context,
+    QueryProcessingStage::Enum processed_stage,
+    size_t max_block_size,
+    unsigned num_streams)
+{
+    StorageS3::updateS3Configuration(context, base_configuration);
+
+    return s3engine->read(column_names, storage_snapshot, query_info, context, processed_stage, max_block_size, num_streams);
+}
+
+String StorageDelta::generateQueryFromKeys(std::vector<String> && keys)
+{
+    std::string new_query = fmt::format("{{{}}}", fmt::join(keys, ","));
```

Comment:
so it will be `{{key1, key2, ...}}` ? why this format?

Reply:
Delta Lake stores table data split in parts, keys is filenames of parts. So, for StorageS3 to read all parts of table we need query ```{key1, key2, ...}```. But '{' is special character in fmt::format, so it needs to be escaped with another '{'.